cirq-core/cirq/devices/noise_utils.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Sequence
+import warnings
+import numpy as np
+
+if TYPE_CHECKING:
+    import cirq
+
+
+# Tag for gates to which noise must be applied.
+PHYSICAL_GATE_TAG = 'physical_gate'
+
+
+@dataclass(frozen=True)
+class OpIdentifier:
```

Comment:
Is this class really needed ? (having looked at other PRs as well).

Reply:
We could _probably_ get away without it, but the result would be a lot messier. Here are the alternatives I considered:

- Just use `Operation` (it's hashable and contains the required info): this requires every current `OpIdentifier` reference to have some concept of "canonical" gates for each type supported in noise model generation (e.g. the `PhasedXZGate` constructor requires `(x|z|axis_phase)_exponent` args). This is considerably less convenient than just converting OpIDs to Operations during serialization.
- Use a (Gate, qubits) tuple: this works about the same as `OpIdentifier`, but it's ultimately less readable to be saying `op_id[0]` and `op_id[1:]` instead of `op_id.gate_type` and `op_id.qubits`

If you have any other recommendations, I'm happy to consider them, but my sense from working through this is that `OpIdentifier` adds readability for relatively little cost. I opted not to surface `OpIdentifier` at the `cirq` package level to avoid any potential confusion.